display projects list via rails route

davidsbailey opened the pull request:
In the previous PR https://github.com/code-dot-org/code-dot-org/pull/12689 I created a projects list component in react. This PR exposes the projects list via rails including fetching all the students and projects metadata for the given section from the DB. This dashboard page will then be embedded via iframe into teacher dashboard. A lot of the advanced features in this grid are being deferred until after we've verified that the iframe approach works. For more background info, see [spec](https://docs.google.com/document/d/1fspDmw78LQwkDHsT7vLZMojKeevn0a9pOYJ73_wcc7o/edit#heading=h.zbgdpwn3k997) and [dev plan](https://docs.google.com/document/d/1s0mXgpoHyvpslvW63-XTCKP_jRdqx019IhbMFs7Wtug/edit#).

screenshot:
<img width="638" alt="screen shot 2017-01-18 at 2 37 07 pm" src="https://cloud.githubusercontent.com/assets/8001765/22089326/34444898-dd9f-11e6-91b3-ac3711122a04.png">

davidsbailey commented:
@aoby and @Bjvanminnen , can you please review? Thanks!

davidsbailey commented:
@aoby @Bjvanminnen bumping this to the top of your inboxes, in case it got lost in the evening's mail :)

Bjvanminnen commented on apps/script/checkEntryPoints.js:
```diff
@@ -84,7 +78,7 @@ const SILENCED = [
   'pd',
   'plc',
-  'publicKeyCryptography',
+  'projectsList',
```
From the comments above:

  * You may be tempted to add your entry point to the list of silenced entry  | |
  * | |  points, but that would be a mistake:

davidsbailey replied:
Admittedly I only tried to work around this by looking at how a couple other examples dealt with this, and they all dealt with it by having an entry in this list. I will try harder to avoid adding to this list. If you know of an example that avoids this, please let me know.

Bjvanminnen replied:
Which rule were you violating/getting a warning about/silencing?

davidsbailey replied:
I could have sworn it was a confusing error message the first time but when I look again it seems pretty clear:
```
✘ projectsList ➜ ./src/sites/studio/pages/projectsList.js
  ✘ Entry point names should match the name of the file they are used in.
    This entry point should be renamed to projects_lists/index!
  this entry point is referenced by the following templates:
    ../dashboard/app/views/projects_lists/index.html.haml
```

Bjvanminnen commented on apps/src/templates/projects/ProjectsList.jsx:
```diff
@@ -133,3 +136,6 @@ const ProjectsList = React.createClass({
 export default ProjectsList;
+
+window.dashboard = window.dashboard || {};
+window.dashboard.ProjectsList = ProjectsList;
```
Would be better if this happened in the entry file (apps/src/sites/studio/pages/projectsList.js) instead of here.

Bjvanminnen commented on dashboard/app/controllers/projects_lists_controller.rb:
```diff
@@ -0,0 +1,41 @@
+class ProjectsListsController < ApplicationController
+  include LevelsHelper
+
+  def index
```
Could use a comment saying what this method does.

Bjvanminnen commented on dashboard/app/controllers/projects_lists_controller.rb:
```diff
@@ -0,0 +1,41 @@
+      updatedAt: project_value['updatedAt'],
+    }
+  end
+end
```
Could use some unit tests.

Bjvanminnen commented on dashboard/app/controllers/projects_lists_controller.rb:
```diff
@@ -0,0 +1,41 @@
+    level && level.split('/')[2]
+  end
+
+  def get_project_row_data(student, project, channel_id)
```
Would be great to have a function comment, explaining among other things what types we expect these params to be.

aoby commented on dashboard/app/controllers/projects_lists_controller.rb:
```diff
@@ -0,0 +1,41 @@
+  def index
+    section_id = params[:section_id]
+    section_owner_id = Section.find(section_id).user_id
+    raise "section can only be accessed by its owner" unless current_user && current_user.id == section_owner_id
```
Leverage `CanCan` here instead of checking authorization explicitly. We already have this rule for sections in [ability](https://github.com/code-dot-org/code-dot-org/blob/0dfac548449075832d9acdda445c83f0b807034b/dashboard/app/models/ability.rb#L62-L62). You should be able to load the appropriate section (and raise the appropriate error implicitly) via `load_and_authorize_resource` and remove these 3 lines.

aoby replied:
Please add unit tests to ensure authorization and basic functionality.

davidsbailey replied:
Hey @aoby, the CanCan step seems to work when I add `load_and_authorize_resource :section`, but only when the placeholder in the route is named `:section_id`, i.e. `get "/:section_id", to: 'projects_lists#index'`. Why is this? I don't see any hints in the documentation at https://github.com/ryanb/cancan/wiki/authorizing-controller-actions, or http://www.rubydoc.info/github/ryanb/cancan/CanCan/ControllerAdditions/ClassMethods, or http://guides.rubyonrails.org/routing.html.

aoby commented on dashboard/app/controllers/projects_lists_controller.rb:
```diff
@@ -0,0 +1,41 @@
+class ProjectsListsController < ApplicationController
```
nit: `ProjectsListsController` implies to me that it will serve lists of projects (meaning `index` will serve multiple lists). Instead it seems to be serving a single list of projects associated with a given section. How about something like `SectionProjectsController`?

davidsbailey replied:
Let me answer your question by asking a question :) in the future, we will want to serve a list of projects that represents all the projects owned by a single student. this will be viewed by the student, not the teacher. should this controller be named so as to accommodate those requests also? or should that be handled by a different controller?

aoby replied:
In that case, the `index` action doesn't seem appropriate. If `index` is all projects for a section, what will we call all projects for a student? 

You have several choices:
 1. You could overload `index` and filter with params such that it returns all projects for a section and/or student. 
 2. You could rename the actions, instead of a generic index, provide 2 appropriate actions
 3. 2 controllers as you suggested

This also brings up another question. Why does this need to be its own controller instead of new actions on the existing `ProjectsController`? Are you separating it because it's a different view and different context?

davidsbailey replied:
Good question, this should probably really go in the projects controller, especially since the students view of their own projects will replace what you see today at `/projects/`, which is currently served from the `ProjectsController`. I'll add this to the `ProjectsController` as `/projects/section/:id` or something like that.

aoby commented on dashboard/app/controllers/projects_lists_controller.rb:
```diff
@@ -0,0 +1,41 @@
+    raise "section can only be accessed by its owner" unless current_user && current_user.id == section_owner_id
+
+    @projects_list_data = []
+    Section.find(section_id).students.each do |student|
```
nit: this `Section.find` would have been a 2nd DB call (duplicate with line 6). Better to grab the section once and reuse it.

aoby commented on dashboard/app/controllers/projects_lists_controller.rb:
```diff
@@ -0,0 +1,41 @@
+      type: project_type(project_value['level']),
+      updatedAt: project_value['updatedAt'],
+    }
+  end
```
This is a bit of logic in the controller. Is this duplicated somewhere else? Can we consolidate into a model or helper?

davidsbailey replied:
No, this logic is not duplicated anywhere else, but I'm fine with moving it. Please advise on whether I should move it to a helper or a model.

a few thoughts:
* since we're not persisting project lists to the db, we might not need a model
* using a model, I could to pull in some values "through" other models such as users/students and sections rather that using explicit ActiveRecord calls. caveat: I won't be able to use activerecord to access `user_storage_ids` or `storage_apps` / projects since those are in pegasus.

aoby replied:
Actually now that I look at it, it seems to be mostly formatting the results into JSON, not really performing much logic. It's probably best to put this in a `Serializer`. I'm also ok with it being here, but please add a comment like @Bjvanminnen suggested.

davidsbailey replied:
per offline discussion, I'll add something in `dashboard/lib` that fetches the data from pegasus, similar to https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/lib/api/v1/pd/workshop_score_summarizer.rb#L83 .

davidsbailey replied:
Reading about this more, especially [this blog post](http://codefol.io/posts/where-do-i-put-my-code-in-rails-updated), it seems like this logic belongs in the ~~controller~~ [correction: model] or lib, and that we're heavily misusing helpers by including them from ~~models~~ (correction: controllers). from the above post:
* With **models**, also remember that there’s no requirement that they be based on ActiveRecord. It’s fine for them to use a different data store... It’s also fine for them to only exist in memory.
* **Rails “helpers”** are very specifically view helpers. They’re automatically included in views, but not in controllers or models. That’s on purpose.
* Do you need something like a “controller helper” or “model helper”? You can put it in a parent class if that makes sense. Or you can put it into **the /lib directory**...

aoby replied:
Don't you mean *model* or lib? Still, after reading that blog post, I contend this logic ideally does not belong in the controller. `"In general, the controller should be a tiny, thin glue layer putting together your other components." ` When I suggested "helper" earlier I meant it in the general sense of a utility class (i.e. `lib/`), not specifically rails view helpers, which I agree is the wrong place for this.

aoby replied:
When you say we're heavily misusing helpers by including them in models, can you point to an example? I wasn't aware we were doing that.

davidsbailey replied:
oops, I mixed up models and controllers in my previous response. I've corrected it now.

yes, we should have logic which retrieves data from pegasus in models (not controllers) or libs.

we include helpers from many controllers (not models). `include LevelsHelper`appears in 7 controllers and one model.

aoby replied:
:( You're right. That should not be a helper, if it's intended to be included in models and controllers. It should most likely be a `Concern` if it's cross-cutting logic relevant to one of those layers. If it's more general, it should probably be in `lib`.

aoby commented on dashboard/config/routes.rb:
```diff
@@ -115,6 +115,12 @@ module OPS
+    collection do
+      get "/:section_id", to: 'projects_lists#index'
+    end
+  end
```
Since this is just one route, you could also define it on a single line as 
```ruby
get '/projects-lists/:section_id', to: 'projects_lists#index'
```

Bjvanminnen commented on dashboard/app/views/projects_lists/index.html.haml:
```diff
@@ -0,0 +1,13 @@
+
+:javascript
+  ReactDOM.render(React.createElement(window.dashboard.ProjectsList, #{reactProps.to_json}),
+    document.getElementById('projects-list'));
```
We've started taking a different approach to avoid using window.dashboard.ProjectList for things like this. Instead, we pass in the necessary data as a data-attribute on the script. As an example
https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/sites/studio/pages/scriptOverview.js
https://github.com/code-dot-org/code-dot-org/blob/47ebd54cfa96ebdc9a04ba74fdc61f61bd2deea6/dashboard/app/views/scripts/show.html.haml#L76

davidsbailey replied:
👍 thanks for pointing this out, I like the data attributes approach much better!

davidsbailey commented:
@aoby @Bjvanminnen could you please both take another look?

I've addressed all feedback with hopefully useful individual commit descriptions. One caveat is that I'm proposing merging this as a rails route without knowing for sure that our strategy (TBD) for getting this into teacher dashboard will require it. The options are:
1. embed `/projects/sections/:section_id` as an iframe
2. serve the JS file through pegasus and remove the `/projects/sections/:section_id` route

If we go with option (2), I'm not too worried about code churn since there are only about 10 lines of code associated with the rails route that would be going away, but I'm open to holding off on merging this until the above decision is made if you guys think that is better.

aoby commented on dashboard/config/routes.rb:
```diff
@@ -115,6 +115,8 @@ module OPS
     end
   end
 
+  get '/projects/section/:section_id', to: 'projects#section_projects'
```
Since this is now using the existing `ProjectsController`, if this route is to stay in Rails, it should go with the other projects routes (add the following line to the [collection](https://github.com/code-dot-org/code-dot-org/blob/c2d98235695272ff5d389afa93209207c01faaa3/dashboard/config/routes.rb#L104-L104) collection just above):

```ruby
get 'section/:section_id', action: 'section_projects'
```

Bjvanminnen commented on apps/script/checkEntryPoints.js:
```diff
@@ -44,12 +44,6 @@ const path = require('path');
-  'makerlab',
-  'pd',
-  'publicKeyCryptography',
-  'brambleHost',
```
Why are we deleting all these? Were they fixed?

davidsbailey replied:
These are not fixed, I am just removing some duplicates from this list.

Bjvanminnen commented on apps/src/sites/studio/pages/projects/section_projects.js:
```diff
@@ -0,0 +1,27 @@
+    url: dataUrl,
+    dataType: 'json'
+  }).done(projectsData => {
+    ReactDOM.render(React.createElement(ProjectsList, {projectsData}), element);
```
Can just do:
`ReactDOM.render(<ProjectsList ...projectsData}/>, element);

davidsbailey replied:
The closest I can get to that is `ReactDOM.render(<ProjectsList projectsData={projectsData}/>, element);`.

Bjvanminnen commented on dashboard/app/controllers/projects_controller.rb:
```diff
@@ -157,6 +157,9 @@ def set_level
   end
 
+  def section_projects
+  end
```
What does this accomplish? Is it just so that rails does rails magic and renders section_projects.html.haml? Could/should we make that explicit here for better understanding? Might also be worth a function comment saying what section_projects action is.

davidsbailey replied:
I'm not sure if the empty function is necessary here, but it turns out I needed to add some code here anyway to hide the header and footer. I've also added a comment.

Bjvanminnen commented on dashboard/lib/projects_list.rb:
```diff
@@ -0,0 +1,41 @@
+module ProjectsList
+  class << self
+    def fetch_section_projects(section)
```
Could you add a comment for this method?

aoby replied:
Please document this public function, ideally in [YARD](http://www.rubydoc.info/gems/yard/file/docs/GettingStarted.md#Which_Markup_Format_) syntax, especially since in `lib` it may not be entirely obvious what section is.

Bjvanminnen commented on dashboard/lib/projects_list.rb:
```diff
@@ -0,0 +1,41 @@
+    # pull various fields out of the student and project records to populate
+    # a data structure that can be used to populate a UI component displaying a
+    # list of projects.
+    def get_project_row_data(student, project, channel_id)
```
worth adding one or more unit tests for this method?

davidsbailey replied:
I've added some unit tests for `get_project_row_data `.

Bjvanminnen commented:
some minor comments, but generally looks good :)

aoby commented on dashboard/config/routes.rb:
```diff
@@ -395,6 +397,10 @@ module OPS
+
+      namespace :projects do
+        get 'section/:section_id', to: 'section_projects#index', defaults: { format: 'json' }
+      end
```
:( Outside the scope of this PR, but I wonder why this `api/v1` namespace is defined in multiple places. We should consolidate.

[line 330](https://github.com/code-dot-org/code-dot-org/blob/c2d98235695272ff5d389afa93209207c01faaa3/dashboard/config/routes.rb#L330-L334)
and [line 389](https://github.com/code-dot-org/code-dot-org/blob/c2d98235695272ff5d389afa93209207c01faaa3/dashboard/config/routes.rb#L389-L399)

I won't mind if you do this as part of this PR ;), but no need.

aoby commented on dashboard/lib/projects_list.rb:
```diff
@@ -0,0 +1,41 @@
+module ProjectsList
+  class << self
+    def fetch_section_projects(section)
```
style: we prefer `def self.fetch_section_projects(section)` over `class << self` See [style guide](https://github.com/bbatsov/ruby-style-guide#def-self-class-methods).

Along these lines, 
`class << self` shows up [twice](https://github.com/code-dot-org/code-dot-org/search?utf8=%E2%9C%93&q=%22class+%3C%3C+self%22&type=Code) in our codebase vs. [193](https://github.com/code-dot-org/code-dot-org/search?utf8=%E2%9C%93&q=%22def+self.%22&type=Code) for `def self.`

davidsbailey replied:
Ok, but now I'm using `private_class_method` to keep those methods private, and the only other examples of that keyword are ones that I introduced.

aoby replied:
Ah I didn't notice the privates were also class methods. In that case I don't care strongly, especially since the style guide does allow `class << self`

aoby commented on dashboard/test/controllers/api/v1/projects/section_projects_controller_test.rb:
```diff
@@ -0,0 +1,55 @@
+    student_project = {id: 22, value: student_project_value}
+
+    hidden_project_value = '{"name":"Hidden App","level":"/projects/playlab","createdAt":"2017-01-01T00:00:00.000-08:00",'\
+      '"updatedAt":"2017-01-01T00:00:00.000-08:00","hidden":true}'
```
not super important, but since this is spanning multiple rows anyway it might be more readable to split over multiple lines. You can even leverage Ruby's hash/JSON support like this:

```ruby
hidden_project_value = {
  name: 'Hidden App',
  level: '/projects/playlab',
  createdAt: '2017-01-01T00:00:00.000-08:00',
  updatedAt: '2017-01-01T00:00:00.000-08:00',
  hidden: true
}.to_json
```

aoby commented on dashboard/test/controllers/api/v1/projects/section_projects_controller_test.rb:
```diff
@@ -0,0 +1,55 @@
+  test 'student cannot access section projects' do
+    sign_in(@student)
+    get :index, params: {section_id: @section.id}
+    assert_equal @response.status, 403
```
Can also use `assert_response 403` or `assert_response :forbidden`

aoby commented on dashboard/test/controllers/api/v1/projects/section_projects_controller_test.rb:
```diff
@@ -0,0 +1,55 @@
+    sign_in(@teacher)
+    get :index, params: {section_id: @section.id}
+    assert_equal @response.status, 200
+  end
```
Add a test to verify a teacher can only view his own sections, i.e. attempting to view a section with another owner gives 403 (or maybe 404?)

aoby replied:
Also (according to `ability.rb`) admins can read sections. I'm not sure if this is functionality we care about, but if it is, it might be worth adding a unit test ensuring that admins can view sections (even those that don't belong to them).

davidsbailey commented on dashboard/test/lib/projects_list_test.rb:
```diff
@@ -0,0 +1,30 @@
+      '"updatedAt":"2017-01-25T17:48:12.358-08:00"}'
+    @student_project = {id: 22, value: student_project_value}
+
+    hidden_project_value = '{"name":"Hidden App","level":"/projects/playlab","createdAt":"2017-01-01T00:00:00.000-08:00",'\
```
@aoby do you think these are a good candidate for factorygirl, or is that only appropriate for activerecord objects?

aoby replied:
You can construct pretty much anything there, but it's generally more useful for reusable models. These look like arbitrary JSON strings, that have no dependencies and are easy enough to construct directly. Your call, but I'd lean toward leaving them inline in the test.

davidsbailey commented:
Thanks @aoby and @Bjvanminnen , I've addressed all of your feedback. Please let me know if you have any additional comments, otherwise I'll go ahead and merge once unit tests are passing on CircleCI.
